scripts/each-service.js

```diff
@@ -0,0 +1,118 @@
+// Helper script that can read the set of services included
+// in the azure SDK dynamically.
+
+var fs = require('fs');
+var path = require('path');
+var util = require('util');
+
+var sdkRoot = path.join(__dirname, '..');
+var servicesRoot = path.join(sdkRoot, 'lib/services');
+
+//
+// Can't use readable-stream, have to only use core modules
+// since this is called by ci before npm install runs.
+//
+
+//
+// Implementation of async versions of fundamental higher-order functions
+// over arrays.
+//
+function reduceAsync(items, asyncIterator, done, initialValue) {
+  if (!initialValue) { initialValue = []; }
+  function doReduce(accumulator, remainingItems, index) {
+    if (remainingItems.length === 0) {
+      return done(null, accumulator);
+    }
+
+    asyncIterator(accumulator, remainingItems[0],
+      function (err, newAccumulator) {
+        if (err) { return done(err); }
+        doReduce(newAccumulator, remainingItems.slice(1), index + 1);
+      }, index, items);
+  }
+  doReduce(initialValue, items, 0);
+}
+
+function filterAsync(items, asyncPredicate, done) {
+  reduceAsync(items,
+    function(acc, item, next) {
+      asyncPredicate(item, function(err, passed) {
+          if (err) { return next(err); }
+          if (passed) {
+            acc.push(item);
+          }
+          return next(null, acc);
+        });
+    }, done);
+}
+
+function mapAsync(items, asyncIterator, done) {
+  reduceAsync(items,
+    function (acc, item, next, index, array) {
+      asyncIterator(item,
+        function (err, mappedItem) {
+          if (err) { return next(err); }
+          return next(null, acc.concat(mappedItem));
+        },
+        index, array);
+    }, done);
+}
+
+function forEachAsync(items, asyncIterator, done) {
+  reduceAsync(items,
+    function(acc, item, next, index, array) {
+      asyncIterator(item,
+        function (err) { return next(err, acc); },
+        index, array);
+    },
+    function (err, result) {
+      done(err);
+    }
+  );
+}
+
+//
+// Find the list of potential service directories, including common
+//
+function readServiceDirs(done) {
+  fs.readdir(servicesRoot, function (err, files) {
+    if (err) { return done(err); }
+    done(null, [path.join(sdkRoot, 'lib/common')].concat(
+      files.map(function (f) {
+        return path.join(servicesRoot, f);
+      })));
+  });
+}
+
+function hasPackageJson(servicePath, done) {
```

Comment:
This routine appears just to check the service path exists, rather check the package.json exists. I think we should add code to check the file exists. Checking the folder might not be meaningful as the collection were returned from  "fs.readdir".